File: packages/qgis-sdk/src/qgis_sdk/bridge/qgis_api/network.py
"""qgis_sdk.bridge.qgis_api.network — network via QGIS NAM for JS."""

from __future__ import annotations

from typing import Any, Dict, Optional


class NetworkAPI:
    def fetch(self, url: str, method: str = "GET", headers: Dict[str, str] = None, body: str = None, auth_cfg: str = None) -> Dict[str, Any]:
        """Makes network request via QgsNetworkAccessManager (respects QGIS proxy, auth).

        Returns dict with status, headers, body, ok.
        """
        headers = headers or {}
        try:
            from qgis_sdk.network import Session
            sess = Session(auth_cfg=auth_cfg, headers=headers)
            # body may be string, convert
            data = body
            resp = sess.request(method, url, data=data, headers=headers, auth_cfg=auth_cfg)
            return {
                "status": resp.status_code,
                "headers": dict(resp.headers),
                "body": resp.text,
                "ok": resp.ok,
                "url": resp.url,
            }
        except Exception as e:
            # Fallback urllib
            try:
                import urllib.request
                req = urllib.request.Request(url, data=body.encode() if isinstance(body, str) else body, headers=headers or {}, method=method)
                with urllib.request.urlopen(req, timeout=15) as r:
                    content = r.read().decode('utf-8', errors='ignore')
                    return {
                        "status": r.getcode(),
                        "headers": dict(r.getheaders()),
                        "body": content,
                        "ok": 200 <= r.getcode() < 400,
                        "url": url,
                    }
            except Exception as e2:
                return {
                    "status": 0,
                    "headers": {},
                    "body": "",
                    "ok": False,
                    "error": str(e2),
                    "url": url,
                }

    def get(self, url: str, headers: Dict[str, str] = None, auth_cfg: str = None) -> Dict[str, Any]:
        return self.fetch(url, method="GET", headers=headers, auth_cfg=auth_cfg)

    def post(self, url: str, body: str = None, headers: Dict[str, str] = None, auth_cfg: str = None) -> Dict[str, Any]:
        return self.fetch(url, method="POST", headers=headers, body=body, auth_cfg=auth_cfg)

File: packages-bun/qgis-sdk-bridge/src/qgis/network.ts
/** qgis.network API — via QGIS NAM */

import type { QgisBridge } from '../window';

export interface QgisNetworkResponse {
  ok: boolean;
  status: number;
  headers: Record<string, string>;
  url: string;
  body?: string;
  error?: string;
  text(): Promise<string>;
  json(): Promise<any>;
}

export class NetworkAPI {
  constructor(private _bridge: QgisBridge, private _raw: any) {}

  private async _call(method: string, ...args: any[]): Promise<any> {
    if (this._raw && typeof this._raw[method] === 'function') {
      return new Promise((resolve, reject) => {
        try {
          this._raw[method](...args, (res: any) => {
            if (typeof res === 'string') {
              try {
                const parsed = JSON.parse(res);
                resolve(parsed);
                return;
              } catch {}
            }
            resolve(res);
          });
        } catch (e) {
          reject(e);
        }
      });
    }
    if (this._bridge && typeof (this._bridge as any)[method] === 'function') {
      return (this._bridge as any)[method](...args);
    }
    // Fallback to browser fetch if not in QGIS (for testing)
    console.warn(`[qgis.network] ${method} fallback to browser fetch`);
    const url = args[0];
    try {
      const resp = await fetch(url);
      const text = await resp.text();
      return {
        status: resp.status,
        headers: Object.fromEntries(resp.headers.entries()),
        body: text,
        ok: resp.ok,
        url,
      };
    } catch (e) {
      return { status: 0, headers: {}, body: '', ok: false, error: String(e), url };
    }
  }

  async fetch(url: string, opts: { method?: string; headers?: any; body?: string; authCfg?: string } = {}): Promise<QgisNetworkResponse> {
    const res = await this._call('network_fetch', url, opts.method || 'GET', opts.headers || {}, opts.body || null, opts.authCfg || null);
    
    return {
      ok: res.ok,
      status: res.status,
      headers: res.headers || {},
      url: res.url || url,
      body: res.body,
      error: res.error,
      text: async () => res.body || '',
      json: async () => {
        try {
          return JSON.parse(res.body || '{}');
        } catch {
          return null;
        }
      }
    };
  }

  async get(url: string, opts: { headers?: any; authCfg?: string } = {}): Promise<QgisNetworkResponse> {
    return this.fetch(url, { method: 'GET', ...opts });
  }

  async post(url: string, body?: string, opts: { headers?: any; authCfg?: string } = {}): Promise<QgisNetworkResponse> {
    return this.fetch(url, { method: 'POST', body, ...opts });
  }
}
